felix/bpf/bpfmap/bpf_maps.go

```diff
@@ -55,25 +55,20 @@ type Maps struct {
 	XDPJumpMap      maps.MapWithDeleteIfExists
 }
 
-func (m *Maps) Destroy() {
-	mps := []maps.Map{
-		m.IpsetsMap,
-		m.StateMap,
-		m.ArpMap,
-		m.FailsafesMap,
-		m.FrontendMap,
-		m.BackendMap,
-		m.AffinityMap,
-		m.RouteMap,
-		m.CtMap,
-		m.SrMsgMap,
-		m.CtNatsMap,
-		m.ProgramsMap,
-		m.JumpMap,
-		m.XDPProgramsMap,
-		m.XDPJumpMap,
+func (m *Maps) DeletePreviousVersion() error {
+	mps := m.extractMapSlice()
+	for _, m := range mps {
+		err := m.(pinnedMap).DeletePreviousVersion()
+		if err != nil {
+			return err
+		}
```

Comment:
I was kinds hoping for something simpler that would also remove pins if we skipped a version, which would make the cleanup a bit more robust without adding this much boilerplate code to the maps. Unfortunately I think we also mix some program pins in this dir. But perhaps not and that would make the code simpler.

```suggestion
expectedPins[m.GetName()] = true
filepath.WalkDir(tcdefs.GlobalPinDir, func(f) {
    if f not in expectedPins and f not prefix of anything in expectedPins {
       delete f
    }
})
```

Reply:
Update: actually further grok'ing in the private helper method [getOldMapVersion()](https://github.com/projectcalico/calico/blob/master/felix/bpf/maps/maps.go#L778)  seems to assume that any previous version of a pinned map will be located in the same  `pinDir()` of the current version from this LOC:
```
files, err := os.ReadDir(b.pinDir())
```
which may also not be reliable / robust.